server/src/main/scala/controllers/Application.scala

```diff
@@ -424,6 +428,63 @@ class Application @Inject()(implicit val config: Configuration,
     Ok(Json.obj("reload" -> shouldRedirect))
   }
 
+  def apiLogin(): Action[Map[String, Seq[String]]] = Action.async(parse.tolerantFormUrlEncoded) { request =>
+
+    def postStringValOrElse(key: String): Option[String] = {
+      request.body.get(key).map(_.head)
+    }
+
+    val tokenUrlOption = config.getOptional[String]("key-cloak.token_url")
+    val clientIdOption = config.getOptional[String]("key-cloak.client_id")
+    val clientSecretOption = config.getOptional[String]("key-cloak.client_secret")
+    val usernameOption = postStringValOrElse("username")
+    val passwordOption = postStringValOrElse("password")
+
+    val result: Option[Future[Result]] = for {
+      tokenUrl <- tokenUrlOption
+      clientId <- clientIdOption
+      clientSecret <- clientSecretOption
+    } yield {
+
+      (usernameOption, passwordOption) match {
+        case (Some(username), Some(password)) =>
+          val authClient =  new KeyCloakAuth(tokenUrl, clientId, clientSecret, system) with ProdSendAndReceive
+          authClient.getToken(username, password).map(token => {
```

Comment:
could extract this closure to a function taking a token to make the match easier to read